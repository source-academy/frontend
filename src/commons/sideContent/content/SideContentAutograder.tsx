import { Button, Collapse, Icon, PopoverPosition, Tooltip } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { useCallback, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';

import { AutogradingResult, Testcase } from '../../assessment/AssessmentTypes';
import ControlButton from '../../ControlButton';
import { WorkspaceLocation } from '../../workspace/WorkspaceTypes';
import SideContentResultCard from './SideContentResultCard';
import SideContentTestcaseCard from './SideContentTestcaseCard';

export type SideContentAutograderProps = DispatchProps & StateProps & OwnProps;

type DispatchProps = {
  handleTestcaseEval: (testcaseId: number) => void;
};

type StateProps = {
  autogradingResults: AutogradingResult[];
  testcases: Testcase[];
};

type OwnProps = {
  /**
   * We need to know the workspace location to hide 'opaque' testcases
   * in AssessmentsWorkspace, but show them in GradingWorkspace.
   */
  workspaceLocation: WorkspaceLocation;
};

const SideContentAutograder: React.FC<SideContentAutograderProps> = props => {
  const { t } = useTranslation('sideContent', { keyPrefix: 'autograder' });
  const [showsTestcases, setTestcasesShown] = useState(true);
  const [showsResults, setResultsShown] = useState(true);

  const { testcases, autogradingResults, handleTestcaseEval, workspaceLocation } = props;

  const autograderTooltip = useMemo(
    () => (
      <div className="autograder-help-tooltip">
        <p>{t('tooltip.clickTestcase')}</p>
        <p>{t('tooltip.executeAll')}</p>
        <p>{t('tooltip.backgroundInfo')}</p>
        <p>{t('tooltip.privateTestcases')}</p>
      </div>
    ),
    [t]
  );

  const testcasesHeader = useMemo(
    () => (
      <div className="testcases-header" data-testid="testcases-header">
        {columnHeader('header-fn', t('headers.testcase'))}
        {columnHeader('header-expected', t('headers.expected'))}
        {columnHeader('header-actual', t('headers.actual'))}
      </div>
    ),
    [t]
  );

  const resultsHeader = useMemo(
    () => (
      <div className="results-header" data-testid="results-header">
        <div className="header-data">
          {columnHeader('header-sn', t('headers.sn'))}
          {columnHeader('header-status', t('headers.status'))}
        </div>
        {columnHeader('header-expected', t('headers.expected'))}
        {columnHeader('header-actual', t('headers.actual'))}
      </div>
    ),
    [t]
  );

  const testcaseCards = useMemo(
    () =>
      testcases.length > 0 ? (
        <div className="testcaseCards">
          {testcasesHeader}
          {testcases.map((testcase, index) => (
            <SideContentTestcaseCard
              key={index}
              index={index}
              testcase={testcase}
              handleTestcaseEval={handleTestcaseEval}
              workspaceLocation={workspaceLocation}
            />
          ))}
        </div>
      ) : (
        <div className="noResults" data-testid="noResults">
          {t('noTestcases')}
        </div>
      ),
    [testcases, testcasesHeader, t, handleTestcaseEval, workspaceLocation]
  );

  const resultCards = useMemo(
    () =>
      autogradingResults.length > 0 ? (
        <div>
          {resultsHeader}
          {autogradingResults.map((result, index) => (
            <SideContentResultCard key={index} index={index} result={result} />
          ))}
        </div>
      ) : (
        <div className="noResults" data-testid="noResults">
          {t('noResults')}
        </div>
      ),
    [autogradingResults, resultsHeader, t]
  );

  const toggleTestcases = useCallback(() => {
    setTestcasesShown(!showsTestcases);
  }, [showsTestcases]);

  const toggleResults = useCallback(() => setResultsShown(!showsResults), [showsResults]);

  return (
    <div className="Autograder">
      <Button
        className="collapse-button"
        icon={showsTestcases ? IconNames.CARET_DOWN : IconNames.CARET_RIGHT}
        minimal={true}
        onClick={toggleTestcases}
      >
        <span>{t('testcases')}</span>
        <Tooltip content={autograderTooltip} placement={PopoverPosition.LEFT}>
          <Icon icon={IconNames.HELP} />
        </Tooltip>
      </Button>
      <Collapse isOpen={showsTestcases} keepChildrenMounted={true}>
        {testcaseCards}
      </Collapse>
      {collapseButton(t('results'), showsResults, toggleResults)}
      <Collapse isOpen={showsResults} keepChildrenMounted={true}>
        {resultCards}
      </Collapse>
    </div>
  );
};

const columnHeader = (colClass: string, colTitle: string) => (
  <div className={colClass}>
    {colTitle}
    <Icon icon={IconNames.CARET_DOWN} />
  </div>
);

const collapseButton = (label: string, isOpen: boolean, toggleFunc: () => void) => (
  <ControlButton
    label={label}
    icon={isOpen ? IconNames.CARET_DOWN : IconNames.CARET_RIGHT}
    onClick={toggleFunc}
    options={{ className: 'collapse-button', minimal: true }}
  />
);

export default SideContentAutograder;
