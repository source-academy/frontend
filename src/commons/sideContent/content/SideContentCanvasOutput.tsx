import React, { useEffect, useRef } from 'react';

type SideContentCanvasOutputProps = StateProps;

type StateProps = {
  canvas: HTMLCanvasElement;
};

/**
 * Takes the output of the rendered graphics (in a hidden canvas tag under <body>)
 * and makes it into a new <canvas> output for viewing.
 */
const SideContentCanvasOutput: React.FC<SideContentCanvasOutputProps> = ({ canvas }) => {
  const parentRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    parentRef.current!.appendChild(canvas);
    canvas.hidden = false;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return <div ref={parentRef} className="canvas-container" />;
};

export default SideContentCanvasOutput;
