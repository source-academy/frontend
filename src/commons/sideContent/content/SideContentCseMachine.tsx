import {
  AnchorButton,
  Button,
  ButtonGroup,
  Checkbox,
  Classes,
  Divider,
  Slider
} from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Tooltip2 } from '@blueprintjs/popover2';
import classNames from 'classnames';
import { debounce } from 'lodash';
import React from 'react';
import { HotKeys } from 'react-hotkeys';
import { connect, MapDispatchToProps, MapStateToProps } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Output } from 'src/commons/repl/Repl';
import type { PlaygroundWorkspaceState } from 'src/commons/workspace/WorkspaceTypes';
import CseMachine from 'src/features/cseMachine/CseMachine';
import { CseAnimation } from 'src/features/cseMachine/CseMachineAnimation';
import { Layout } from 'src/features/cseMachine/CseMachineLayout';

import { InterpreterOutput, OverallState } from '../../application/ApplicationTypes';
import { HighlightedLines } from '../../editor/EditorTypes';
import Constants, { Links } from '../../utils/Constants';
import {
  evalEditor,
  setEditorHighlightedLinesControl,
  updateCurrentStep
} from '../../workspace/WorkspaceActions';
import { beginAlertSideContent } from '../SideContentActions';
import { getLocation } from '../SideContentHelper';
import { NonStoryWorkspaceLocation, SideContentTab, SideContentType } from '../SideContentTypes';

type State = {
  visualization: React.ReactNode;
  value: number;
  height: number;
  width: number;
  lastStep: boolean;
  stepLimitExceeded: boolean;
};

type CseMachineProps = OwnProps & StateProps & DispatchProps;

type StateProps = {
  editorWidth?: string;
  sideContentHeight?: number;
  stepsTotal: number;
  currentStep: number;
  breakpointSteps: number[];
  changepointSteps: number[];
  needCseUpdate: boolean;
  machineOutput: InterpreterOutput[];
};

type OwnProps = {
  workspaceLocation: NonStoryWorkspaceLocation;
};

type DispatchProps = {
  handleStepUpdate: (steps: number) => void;
  handleEditorEval: () => void;
  setEditorHighlightedLines: (
    editorTabIndex: number,
    newHighlightedLines: HighlightedLines[]
  ) => void;
  handleAlertSideContent: () => void;
};

const cseMachineKeyMap = {
  FIRST_STEP: 'a',
  NEXT_STEP: 'f',
  PREVIOUS_STEP: 'b',
  LAST_STEP: 'e'
};

class SideContentCseMachineBase extends React.Component<CseMachineProps, State> {
  constructor(props: CseMachineProps) {
    super(props);
    this.state = {
      visualization: null,
      value: -1,
      width: this.calculateWidth(props.editorWidth),
      height: this.calculateHeight(props.sideContentHeight),
      lastStep: false,
      stepLimitExceeded: false
    };
    CseMachine.init(
      visualization => {
        this.setState({ visualization }, () => CseAnimation.playAnimation());
        if (visualization) this.props.handleAlertSideContent();
      },
      this.state.width,
      this.state.height,
      (segments: [number, number][]) => {
        // TODO: Hardcoded to make use of the first editor tab. Rewrite after editor tabs are added.
        // This comment is copied over from workspace saga
        props.setEditorHighlightedLines(0, segments);
      },
      // We shouldn't be able to move slider to a step number beyond the step limit
      isControlEmpty => {
        this.setState({ stepLimitExceeded: false });
      }
    );
  }

  private calculateWidth(editorWidth?: string) {
    const horizontalPadding = 50;
    const maxWidth = 5000; // limit for visible diagram width for huge screens
    let width;
    if (editorWidth === undefined) {
      width = window.innerWidth - horizontalPadding;
    } else {
      width = Math.min(
        maxWidth,
        (window.innerWidth * (100 - parseFloat(editorWidth))) / 100 - horizontalPadding
      );
    }
    return Math.min(width, maxWidth);
  }

  private calculateHeight(sideContentHeight?: number) {
    const verticalPadding = 150;
    const maxHeight = 5000; // limit for visible diagram height for huge screens
    let height;
    if (window.innerWidth < Constants.mobileBreakpoint) {
      // mobile mode
      height = window.innerHeight - verticalPadding;
    } else if (sideContentHeight === undefined) {
      height = window.innerHeight - verticalPadding;
    } else {
      height = sideContentHeight - verticalPadding;
    }
    return Math.min(height, maxHeight);
  }

  handleResize = debounce(() => {
    const newWidth = this.calculateWidth(this.props.editorWidth);
    const newHeight = this.calculateHeight(this.props.sideContentHeight);
    if (newWidth !== this.state.width || newHeight !== this.state.height) {
      this.setState({
        height: newHeight,
        width: newWidth
      });
      CseMachine.updateDimensions(newWidth, newHeight);
    }
  }, 300);

  componentDidMount() {
    this.handleResize();
    window.addEventListener('resize', this.handleResize);
    CseMachine.redraw();
  }

  componentWillUnmount() {
    this.handleResize.cancel();
    window.removeEventListener('resize', this.handleResize);
  }

  componentDidUpdate(prevProps: {
    editorWidth?: string;
    sideContentHeight?: number;
    stepsTotal: number;
    needCseUpdate: boolean;
  }) {
    if (
      prevProps.sideContentHeight !== this.props.sideContentHeight ||
      prevProps.editorWidth !== this.props.editorWidth
    ) {
      this.handleResize();
    }
    if (prevProps.needCseUpdate && !this.props.needCseUpdate) {
      this.stepFirst();
      CseMachine.clearCse();
    }
  }

  public render() {
    const cseMachineHandlers = this.state.visualization
      ? {
          FIRST_STEP: this.stepFirst,
          NEXT_STEP: this.stepNext,
          PREVIOUS_STEP: this.stepPrevious,
          LAST_STEP: this.stepLast(this.props.stepsTotal)
        }
      : {
          FIRST_STEP: () => {},
          NEXT_STEP: () => {},
          PREVIOUS_STEP: () => {},
          LAST_STEP: () => {}
        };

    return (
      <HotKeys
        keyMap={cseMachineKeyMap}
        handlers={cseMachineHandlers}
        style={{
          maxHeight: '100%',
          overflow: this.state.visualization ? 'hidden' : 'auto'
        }}
      >
        <div className={classNames('sa-substituter', Classes.DARK)}>
          <Slider
            disabled={!this.state.visualization}
            min={0}
            max={this.props.stepsTotal}
            onChange={this.sliderShift}
            onRelease={this.sliderRelease}
            value={this.state.value < 0 ? 0 : this.state.value}
          />
          <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
            <ButtonGroup>
              <Tooltip2 content="Control and Stash" compact>
                <AnchorButton
                  onMouseUp={() => {
                    if (this.state.visualization) {
                      CseMachine.toggleControlStash();
                      CseMachine.redraw();
                    }
                  }}
                  icon="layers"
                  disabled={!this.state.visualization}
                >
                  <Checkbox
                    checked={CseMachine.getControlStash()}
                    disabled={!this.state.visualization}
                    style={{ margin: 0 }}
                  />
                </AnchorButton>
              </Tooltip2>
              <Tooltip2 content="Truncate Control" compact>
                <AnchorButton
                  onMouseUp={() => {
                    if (this.state.visualization) {
                      CseMachine.toggleStackTruncated();
                      CseMachine.redraw();
                    }
                  }}
                  icon="minimize"
                  disabled={!this.state.visualization}
                >
                  <Checkbox
                    checked={CseMachine.getStackTruncated()}
                    disabled={!this.state.visualization}
                    style={{ margin: 0 }}
                  />
                </AnchorButton>
              </Tooltip2>
            </ButtonGroup>
            <ButtonGroup>
              <Button
                disabled={!this.state.visualization}
                icon="double-chevron-left"
                onClick={this.stepPrevBreakpoint}
              />
              <Button
                disabled={!this.state.visualization}
                icon="chevron-left"
                onClick={
                  CseMachine.getControlStash() ? this.stepPrevious : this.stepPrevChangepoint
                }
              />
              <Button
                disabled={!this.state.visualization}
                icon="chevron-right"
                onClick={CseMachine.getControlStash() ? this.stepNext : this.stepNextChangepoint}
              />
              <Button
                disabled={!this.state.visualization}
                icon="double-chevron-right"
                onClick={this.stepNextBreakpoint}
              />
            </ButtonGroup>
            <ButtonGroup>
              <Tooltip2 content="Print" compact>
                <AnchorButton
                  onMouseUp={() => {
                    if (this.state.visualization) {
                      CseMachine.togglePrintableMode();
                      CseMachine.redraw();
                    }
                  }}
                  icon="print"
                  disabled={!this.state.visualization}
                >
                  <Checkbox
                    disabled={!this.state.visualization}
                    checked={CseMachine.getPrintableMode()}
                    style={{ margin: 0 }}
                  />
                </AnchorButton>
              </Tooltip2>
              <Tooltip2 content="Save" compact>
                <AnchorButton
                  icon="floppy-disk"
                  disabled={!this.state.visualization}
                  onClick={Layout.exportImage}
                />
              </Tooltip2>
            </ButtonGroup>
          </div>
        </div>{' '}
        {this.state.visualization &&
        this.props.machineOutput.length &&
        this.props.machineOutput[0].type === 'errors' ? (
          this.props.machineOutput.map((slice, index) => (
            <Output output={slice} key={index} usingSubst={false} isHtml={false} />
          ))
        ) : (
          <div></div>
        )}
        {this.state.visualization ? (
          this.state.stepLimitExceeded ? (
            <div
              id="cse-machine-default-text"
              className={Classes.RUNNING_TEXT}
              data-testid="cse-machine-default-text"
            >
              Maximum number of steps exceeded.
              <Divider />
              Please increase the step limit if you would like to see futher evaluation.
            </div>
          ) : (
            this.state.visualization
          )
        ) : (
          <div
            id="cse-machine-default-text"
            className={Classes.RUNNING_TEXT}
            data-testid="cse-machine-default-text"
          >
            The CSE machine generates control, stash and environment model diagrams following a
            notation introduced in{' '}
            <a href={Links.textbookChapter3_2} rel="noopener noreferrer" target="_blank">
              <i>
                Structure and Interpretation of Computer Programs, JavaScript Edition, Chapter 3,
                Section 2
              </i>
            </a>
            .
            <br />
            <br /> On this tab, the REPL will be hidden from view, so do check that your code has no
            errors before running the stepper. You may use this tool by running your program and
            then dragging the slider above to see the state of the control, stash and environment at
            different stages in the evaluation of your program. Clicking on the fast-forward button
            (double chevron) will take you to the next breakpoint in your program
            <br />
            <br />
            <Divider />
            Some useful keyboard shortcuts:
            <br />
            <br />
            a: Move to the first step
            <br />
            e: Move to the last step
            <br />
            f: Move to the next step
            <br />
            b: Move to the previous step
            <br />
            <br />
            Note that these shortcuts are only active when the browser focus is on this tab.
          </div>
        )}
        <ButtonGroup
          vertical={true}
          style={{ position: 'absolute', bottom: '20px', right: '20px' }}
        >
          <Button
            icon="plus"
            disabled={!this.state.visualization}
            onClick={() => Layout.zoomStage(true, 5)}
            style={{ marginBottom: '5px', borderRadius: '3px' }}
          />
          <Button
            icon="minus"
            disabled={!this.state.visualization}
            onClick={() => Layout.zoomStage(false, 5)}
            style={{ borderRadius: '3px' }}
          />
        </ButtonGroup>
      </HotKeys>
    );
  }

  private sliderRelease = (newValue: number) => {
    if (newValue === this.props.stepsTotal) {
      this.setState({ lastStep: true });
    } else {
      this.setState({ lastStep: false });
    }
    this.props.handleEditorEval();
  };

  private sliderShift = (newValue: number) => {
    this.props.handleStepUpdate(newValue);
    this.setState((state: State) => {
      return { value: newValue };
    });
  };

  private stepPrevious = () => {
    if (this.state.value !== 0) {
      this.sliderShift(this.state.value - 1);
      this.sliderRelease(this.state.value - 1);
    }
  };

  private stepNext = () => {
    const lastStepValue = this.props.stepsTotal;
    if (this.state.value !== lastStepValue) {
      this.sliderShift(this.state.value + 1);
      this.sliderRelease(this.state.value + 1);
      CseAnimation.enableAnimations();
    }
  };

  private stepFirst = () => {
    // Move to the first step
    this.sliderShift(0);
    this.sliderRelease(0);
  };

  private stepLast = (lastStepValue: number) => () => {
    // Move to the last step
    this.sliderShift(lastStepValue);
    this.sliderRelease(lastStepValue);
  };

  private stepNextBreakpoint = () => {
    for (const step of this.props.breakpointSteps) {
      if (step > this.state.value) {
        this.sliderShift(step);
        this.sliderRelease(step);
        return;
      }
    }
    this.sliderShift(this.props.stepsTotal);
    this.sliderRelease(this.props.stepsTotal);
  };

  private stepPrevBreakpoint = () => {
    for (let i = this.props.breakpointSteps.length - 1; i >= 0; i--) {
      const step = this.props.breakpointSteps[i];
      if (step < this.state.value) {
        this.sliderShift(step);
        this.sliderRelease(step);
        return;
      }
    }
    this.sliderShift(0);
    this.sliderRelease(0);
  };

  private stepNextChangepoint = () => {
    for (const step of this.props.changepointSteps) {
      if (step > this.state.value) {
        this.sliderShift(step);
        this.sliderRelease(step);
        return;
      }
    }
    this.sliderShift(this.props.stepsTotal);
    this.sliderRelease(this.props.stepsTotal);
  };

  private stepPrevChangepoint = () => {
    for (let i = this.props.changepointSteps.length - 1; i >= 0; i--) {
      const step = this.props.changepointSteps[i];
      if (step < this.state.value) {
        this.sliderShift(step);
        this.sliderRelease(step);
        return;
      }
    }
    this.sliderShift(0);
    this.sliderRelease(0);
  };
}

const mapStateToProps: MapStateToProps<StateProps, OwnProps, OverallState> = (
  state: OverallState,
  ownProps: OwnProps
) => {
  let workspace: PlaygroundWorkspaceState;
  const [loc] = getLocation(ownProps.workspaceLocation);

  switch (loc) {
    // case 'stories': {
    //   workspace = state.stories.envs[storyEnv]
    //   break
    // }
    case 'sicp': {
      workspace = state.workspaces.sicp;
      break;
    }
    default: {
      workspace = state.workspaces.playground;
      break;
    }
  }

  return {
    ...ownProps,
    stepsTotal: workspace.stepsTotal,
    currentStep: workspace.currentStep,
    breakpointSteps: workspace.breakpointSteps,
    changepointSteps: workspace.changepointSteps,
    needCseUpdate: workspace.updateCse,
    machineOutput: workspace.output
  };
};

const mapDispatchToProps: MapDispatchToProps<DispatchProps, OwnProps> = (dispatch, props) =>
  bindActionCreators(
    {
      handleEditorEval: () => evalEditor(props.workspaceLocation),
      handleStepUpdate: (steps: number) => updateCurrentStep(steps, props.workspaceLocation),
      handleAlertSideContent: () =>
        beginAlertSideContent(SideContentType.cseMachine, props.workspaceLocation),
      setEditorHighlightedLines: (
        editorTabIndex: number,
        newHighlightedLines: HighlightedLines[]
      ) =>
        setEditorHighlightedLinesControl(
          props.workspaceLocation,
          editorTabIndex,
          newHighlightedLines
        )
    },
    dispatch
  );

export const SideContentCseMachine = connect(
  mapStateToProps,
  mapDispatchToProps
)(SideContentCseMachineBase);

const makeCseMachineTabFrom = (location: NonStoryWorkspaceLocation): SideContentTab => ({
  label: 'CSE Machine',
  iconName: IconNames.GLOBE,
  body: <SideContentCseMachine workspaceLocation={location} />,
  id: SideContentType.cseMachine
});

export default makeCseMachineTabFrom;
