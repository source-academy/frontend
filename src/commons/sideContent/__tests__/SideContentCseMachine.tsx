import { act, render, screen } from '@testing-library/react';
import { runInContext } from 'js-slang/dist/';
import { Provider } from 'react-redux';
import { mockInitialStore } from 'src/commons/mocks/StoreMocks';
import { renderTreeJson } from 'src/commons/utils/TestUtils';

import { mockContext } from '../../mocks/ContextMocks';
import { visualizeCseMachine } from '../../utils/JsSlangHelper';
import { SideContentCseMachine } from '../content/SideContentCseMachine';

const mockStore = mockInitialStore();
const element = (
  <Provider store={mockStore}>
    <SideContentCseMachine workspaceLocation="playground" />
  </Provider>
);

test('CSE Machine component renders correctly', () => {
  const tree = renderTreeJson(element);
  expect(tree).toMatchSnapshot();
});

test('CSE Machine sets visualization state and renders', async () => {
  render(element);
  expect(screen.queryAllByTestId('cse-machine-default-text')).toHaveLength(1);
  expect(screen.queryAllByTestId('sa-cse-machine')).toHaveLength(0);

  const context = mockContext();
  runInContext('const hello="world"; debugger;', context);
  act(() => visualizeCseMachine({ context }));

  expect(screen.queryAllByTestId('cse-machine-default-text')).toHaveLength(0);
  expect(screen.queryAllByTestId('sa-cse-machine')).toHaveLength(1);
});
