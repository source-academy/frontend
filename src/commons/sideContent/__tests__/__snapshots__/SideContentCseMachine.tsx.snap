// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSE Machine component renders correctly 1`] = `
<div
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  style={
    Object {
      "maxHeight": "100%",
      "overflow": "auto",
    }
  }
  tabIndex="-1"
>
  <div
    className="sa-substituter bp5-dark"
  >
    <div
      className="bp5-slider bp5-disabled"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="bp5-slider-track"
      >
        <div
          className="bp5-slider-progress bp5-intent-primary"
          style={
            Object {
              "left": "0.00%",
              "right": "100.00%",
              "top": 0,
            }
          }
        />
        <div
          className="bp5-slider-progress"
          style={
            Object {
              "left": "0.00%",
              "right": "100.00%",
              "top": 0,
            }
          }
        />
        <div
          className="bp5-slider-progress"
          style={
            Object {
              "left": "0.00%",
              "right": "100.00%",
              "top": 0,
            }
          }
        />
      </div>
      <div
        className="bp5-slider-axis"
      >
        <div
          className="bp5-slider-label"
          style={
            Object {
              "left": "NaN%",
            }
          }
        >
          0
        </div>
      </div>
      <span
        aria-disabled={true}
        aria-orientation="horizontal"
        aria-valuemax={0}
        aria-valuemin={0}
        aria-valuenow={0}
        className="bp5-slider-handle"
        role="slider"
        style={Object {}}
        tabIndex={0}
      >
        <span
          className="bp5-slider-label"
        >
          0
        </span>
      </span>
    </div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <div
        className="bp5-button-group"
      >
        <span
          className="bp5-popover-target"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            aria-disabled={true}
            aria-expanded={false}
            className="bp5-button bp5-disabled"
            disabled={true}
            onBlur={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseUp={[Function]}
            role="button"
            tabIndex={-1}
          >
            <span
              aria-hidden={true}
              className="bp5-icon bp5-icon-standard bp5-icon-layers"
              data-icon="layers"
            />
            <span
              className="bp5-button-text"
            >
              <label
                className="bp5-control bp5-checkbox bp5-disabled"
                style={
                  Object {
                    "margin": 0,
                  }
                }
              >
                <input
                  checked={false}
                  disabled={true}
                  onChange={[Function]}
                  type="checkbox"
                />
                <span
                  className="bp5-control-indicator"
                />
              </label>
            </span>
          </a>
        </span>
        <span
          className="bp5-popover-target"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            aria-disabled={true}
            aria-expanded={false}
            className="bp5-button bp5-disabled"
            disabled={true}
            onBlur={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseUp={[Function]}
            role="button"
            tabIndex={-1}
          >
            <span
              aria-hidden={true}
              className="bp5-icon bp5-icon-standard bp5-icon-minimize"
              data-icon="minimize"
            />
            <span
              className="bp5-button-text"
            >
              <label
                className="bp5-control bp5-checkbox bp5-disabled"
                style={
                  Object {
                    "margin": 0,
                  }
                }
              >
                <input
                  checked={false}
                  disabled={true}
                  onChange={[Function]}
                  type="checkbox"
                />
                <span
                  className="bp5-control-indicator"
                />
              </label>
            </span>
          </a>
        </span>
      </div>
      <div
        className="bp5-button-group"
      >
        <button
          className="bp5-button bp5-disabled"
          disabled={true}
          onBlur={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            aria-hidden={true}
            className="bp5-icon bp5-icon-standard bp5-icon-double-chevron-left"
            data-icon="double-chevron-left"
          />
        </button>
        <button
          className="bp5-button bp5-disabled"
          disabled={true}
          onBlur={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            aria-hidden={true}
            className="bp5-icon bp5-icon-standard bp5-icon-chevron-left"
            data-icon="chevron-left"
          />
        </button>
        <button
          className="bp5-button bp5-disabled"
          disabled={true}
          onBlur={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            aria-hidden={true}
            className="bp5-icon bp5-icon-standard bp5-icon-chevron-right"
            data-icon="chevron-right"
          />
        </button>
        <button
          className="bp5-button bp5-disabled"
          disabled={true}
          onBlur={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            aria-hidden={true}
            className="bp5-icon bp5-icon-standard bp5-icon-double-chevron-right"
            data-icon="double-chevron-right"
          />
        </button>
      </div>
      <div
        className="bp5-button-group"
      >
        <span
          className="bp5-popover-target"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            aria-disabled={true}
            aria-expanded={false}
            className="bp5-button bp5-disabled"
            disabled={true}
            onBlur={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseUp={[Function]}
            role="button"
            tabIndex={-1}
          >
            <span
              aria-hidden={true}
              className="bp5-icon bp5-icon-standard bp5-icon-print"
              data-icon="print"
            />
            <span
              className="bp5-button-text"
            >
              <label
                className="bp5-control bp5-checkbox bp5-disabled"
                style={
                  Object {
                    "margin": 0,
                  }
                }
              >
                <input
                  checked={false}
                  disabled={true}
                  onChange={[Function]}
                  type="checkbox"
                />
                <span
                  className="bp5-control-indicator"
                />
              </label>
            </span>
          </a>
        </span>
        <span
          className="bp5-popover-target"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            aria-disabled={true}
            aria-expanded={false}
            className="bp5-button bp5-disabled"
            disabled={true}
            onBlur={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            role="button"
            tabIndex={-1}
          >
            <span
              aria-hidden={true}
              className="bp5-icon bp5-icon-standard bp5-icon-floppy-disk"
              data-icon="floppy-disk"
            />
          </a>
        </span>
      </div>
    </div>
  </div>
   
  <div />
  <div
    className="bp5-running-text"
    data-testid="cse-machine-default-text"
    id="cse-machine-default-text"
  >
    The CSE machine generates control, stash and environment model diagrams following a notation introduced in
     
    <a
      href="https://sourceacademy.org/sicpjs/3.2"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i>
        Structure and Interpretation of Computer Programs, JavaScript Edition, Chapter 3, Section 2
      </i>
    </a>
    .
    <br />
    <br />
     On this tab, the REPL will be hidden from view, so do check that your code has no errors before running the stepper. You may use this tool by running your program and then dragging the slider above to see the state of the control, stash and environment at different stages in the evaluation of your program. Clicking on the fast-forward button (double chevron) will take you to the next breakpoint in your program
    <br />
    <br />
    <div
      className="bp5-divider"
    />
    Some useful keyboard shortcuts:
    <br />
    <br />
    a: Move to the first step
    <br />
    e: Move to the last step
    <br />
    f: Move to the next step
    <br />
    b: Move to the previous step
    <br />
    <br />
    Note that these shortcuts are only active when the browser focus is on this tab.
  </div>
  <div
    className="bp5-button-group bp5-vertical"
    style={
      Object {
        "bottom": "20px",
        "position": "absolute",
        "right": "20px",
      }
    }
  >
    <button
      className="bp5-button bp5-disabled"
      disabled={true}
      onBlur={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      style={
        Object {
          "borderRadius": "3px",
          "marginBottom": "5px",
        }
      }
      tabIndex={-1}
      type="button"
    >
      <span
        aria-hidden={true}
        className="bp5-icon bp5-icon-standard bp5-icon-plus"
        data-icon="plus"
      />
    </button>
    <button
      className="bp5-button bp5-disabled"
      disabled={true}
      onBlur={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      style={
        Object {
          "borderRadius": "3px",
        }
      }
      tabIndex={-1}
      type="button"
    >
      <span
        aria-hidden={true}
        className="bp5-icon bp5-icon-standard bp5-icon-minus"
        data-icon="minus"
      />
    </button>
  </div>
</div>
`;
